Add documentation on using virtual environments for testing in napari based on 2022-01 workshop by Talley Lambert (#50)
napari/docs: docs/plugins/virtual_environment_docs/article-2-deploying-your-plugin.md

# Article 2: Tools and tips when building napari plugins: Deploying your plugin  

This article explains some of the techniques you can use to deploy your plugin. 

## Article information  
* Article 1: [Virtual environments](./article-1-virtual-environments.md)   
* Article 2: This article  
* Article 3: [Version management](./article-3-version-management.md)     
* Article 4: [Developer tools](./article-4-developer-tools.md)   
* Article 5: [More than just best practices](./article-5-more-than-just-best-practices.md)  
* Article 6: [Survey](./article-6-Survey.md)   

## This article covers:   
* [PyPI and Anaconda](#pypi-and-anaconda)
    - [Package formats (PyPI)](#package-formats-pypi)
    - [PyPI](#pypi)
* [Deploying plugins](#deploying-plugins)
    - [Automatically via GitHub actions](#automatically-via-github-actions)
    - [Anaconda cloud via conda-forge](#anaconda-cloud-via-conda-forge)

## PyPI and Anaconda

> Try to deploy to both! But for now, try to at least use pip. Currently, the built-in napari plugin installer and the napari-hub, will only look at PyPI. (This is subject to change in the future.) You can always provide your users with manual installation instructions (e.g. if you want to use conda).

Reviewer: We do actually support conda now (or we will very soon) so removing that caveat and slightly tweaking wording.

```suggestion
Try to deploy to both! But for now, try to at least use pip. You can always also provide your users with manual installation instructions (e.g. if you want to use conda or have specific dependencies).
```

### Package formats (PyPI)
**sdist** means source distribution. An **sdist** includes all of the files that are required to *build* your package. An **sdist** may require specific additional software (e.g. compilers) to actually build.

**wheel** is a prebuilt package, ready to drop into site-packages. It includes compiled OS-specific extensions (if applicable).

You are *strongly* encouraged to ship both! If the wheel is not present, pip will try to build it from the sdist before installation, and that may fail depending on the package.  

**Note:** This goes for dependencies too! Check all your dependencies for wheel availability.  

> **Example:** See if a given package ships a wheel. https://pypi.org/project/napari/#files

Reviewer: I'm not really sure what this example is telling me to do/showing me?

Author: I think it is showing you where to check for the wheel?  The example should probably be incorporated into the note.

Author: I changed the text.

### PyPI
**build** (package builder): https://pypa-build.readthedocs.io/en/latest/  
**twine** (PyPI Client): https://twine.readthedocs.io/en/latest/  
```python

    # twine is a PyPI Client  
    # build is a PEP 517 package builder  
    $ pip install twine build

    # create a wheel and an sdist  
    $ python -m build  
     
    # (Optional) upload to test PyPI  
    $ twine upload -r testpypi dist/*  

    # Upload to PyPI  
    $ twine upload dist/*  

```
**Note:** `python -m build` is the modern alternative to setuptools `python setup.py sdist bdist_wheel`. (It calls setuptools behind the scenes.)

## Deploying plugins:

### Automatically via GitHub actions

> This requires either: 

Reviewer: I would just simplify this into one sentence and then introduce the example workflow so:

This requires either running twine manually in a workflow or using a pre-made [GitHub action](https://github.com/pypa/gh-action-pypi-publish). Here is an example workflow that manually deploys using `twine` when tests pass and you push a tagged commit.

Author: I'm not sure where that statement needs to go.  Does it need to go above ### PyPI?
